retrieval/discovery/ec2.go

```diff
@@ -0,0 +1,172 @@
+)
+
+const (
+	ec2Label = model.MetaLabelPrefix + "ec2_"
+)
+
+// EC2Discovery periodically performs EC2-SD requests. It implements
+// the TargetProvider interface.
+type EC2Discovery struct {
+	aws            *aws.Config
+	tags           model.LabelSet
+	regexTags      map[string]*config.Regexp
+	done           chan struct{}
+	interval       time.Duration
+	port           int
+	instanceFilter []*ec2.Filter
+}
+
+// NewEC2Discovery returns a new EC2Discovery which periodically refreshes its targets.
+func NewEC2Discovery(conf *config.EC2SDConfig) *EC2Discovery {
+	creds := credentials.NewStaticCredentials(conf.AccessKey, conf.SecretKey, "")
+	if conf.AccessKey == "" && conf.SecretKey == "" {
+		creds = credentials.NewEnvCredentials()
+	}
+	tags := make(model.LabelSet)
+	filter := &ec2.Filter{
+		Name: aws.String("tag-key"),
+	}
+	for k, v := range conf.Tags {
+		tags[model.LabelName(k)] = model.LabelValue(v.String())
+		filter.Values = append(filter.Values, &k)
+	}
+	return &EC2Discovery{
+		aws: &aws.Config{
+			Region:      &conf.Region,
+			Credentials: creds,
+		},
+		tags:      tags,
+		regexTags: conf.Tags,
+		done:      make(chan struct{}),
+		interval:  time.Duration(conf.RefreshInterval),
+		port:      conf.Port,
+		instanceFilter: []*ec2.Filter{
+			filter,
+		},
+	}
+}
+
+// Run implements the TargetProvider interface.
+func (ed *EC2Discovery) Run(ch chan<- *config.TargetGroup, done <-chan struct{}) {
+	defer close(ch)
+
+	ticker := time.NewTicker(ed.interval)
+	defer ticker.Stop()
+
+	// Get an initial set right away.
+	if err := ed.refresh(ch); err != nil {
+		log.Error(err)
+	}
+
+	for {
+		select {
+		case <-ticker.C:
+			if err := ed.refresh(ch); err != nil {
+				log.Error(err)
+			}
+		case <-done:
+			return
+		}
+	}
+}
+
+// Sources implements the TargetProvider interface.
+func (ed *EC2Discovery) Sources() []string {
+	return []string{*ed.aws.Region}
+}
+
+func (ed *EC2Discovery) refresh(ch chan<- *config.TargetGroup) error {
+	EC2 := ec2.New(ed.aws)
+	dii := ec2.DescribeInstancesInput{
+		Filters: ed.instanceFilter,
+	}
+	tg := &config.TargetGroup{
+		Source: *ed.aws.Region,
+	}
+	if err := EC2.DescribeInstancesPages(&dii, func(p *ec2.DescribeInstancesOutput, lastPage bool) bool {
+		for _, r := range p.Reservations {
+			for _, inst := range r.Instances {
+				labels := model.LabelSet{
+					ec2Label + "instance_id": model.LabelValue(*inst.InstanceId),
+				}
+				if inst.PublicIpAddress != nil {
+					addr := model.LabelValue(fmt.Sprintf("%s:%d", *inst.PublicIpAddress, ed.port))
+					labels[ec2Label+"public_ip"] = addr
+				}
+				if inst.PrivateIpAddress != nil {
+					addr := model.LabelValue(fmt.Sprintf("%s:%d", *inst.PrivateIpAddress, ed.port))
+					labels[ec2Label+"private_ip"] = addr
+					labels[model.AddressLabel] = addr
+				} else {
+					continue
+				}
+				itags := make(map[string]string)
+				for _, t := range inst.Tags {
+					itags[*t.Key] = *t.Value
+					name := cleanTag(*t.Key)
```

Comment:
This was the least bad idea I had about how to fix up the EC2 tags to something valid for prometheus. Is there a better way?